File: packages/mp_dom_runtime/src/engine.ts
declare var require: any;
declare var wx: any;
declare var tt: any;
declare var swan: any;

import { BrowserApp } from "./browser_app";
import { WXApp } from "./wx_app";
import { CustomPaint, MPDrawable } from "./components/basic/custom_paint";
import { WebDialogs } from "./components/basic/web_dialogs";
import { ComponentFactory } from "./components/component_factory";
import { MPPlatformView } from "./components/mpkit/platform_view";
import { MPScaffold } from "./components/mpkit/scaffold";
import { createDebugger, Debugger } from "./debugger/debugger";
import { MPEnv } from "./env";
import { PlatformChannelIO } from "./platform_channel/platform_channel_io";
import { MPJS } from "./mpjs/mpjs";
import { Page } from "./page";
import { Router } from "./router";
import { TextMeasurer } from "./text_measurer";
import { WindowInfo } from "./window_info";
import { wrapDartObject } from "./components/dart_object";
import { CollectionView } from "./components/basic/collection_view";
import { EditableText } from "./components/basic/editable_text";

export class Engine {
  private started: boolean = false;
  private codeBlock?: () => void;
  debugger?: Debugger;
  private messageQueue: string[] = [];
  drawable: MPDrawable;
  componentFactory: ComponentFactory;
  managedViews: { [key: number]: Page } = {};
  unmanagedViewFrameData: { [key: number]: any[] } = {};
  mpJS: MPJS = new MPJS(this);
  app?: BrowserApp | WXApp;
  router?: Router;
  windowInfo = new WindowInfo(this);
  pageMode: boolean = false;
  platformChannelIO: PlatformChannelIO;

  constructor() {
    this.componentFactory = new ComponentFactory(this);
    this.drawable = new MPDrawable(this);
    this.platformChannelIO = new PlatformChannelIO(this);
    this.installWeChatComponentContextGetter();
  }

  private installWeChatComponentContextGetter() {
    if (__MP_MINI_PROGRAM__) {
      MPEnv.platformGlobal().mp_core_weChatComponentContextGetter = async (hashCode: number) => {
        const target = this.componentFactory.cachedView[hashCode];
        if (target) {
          const ctx = await (target.htmlElement as any).$$getContext();
          return ctx;
        }
      };
    }
  }

  public static codeBlockWithCodePath(codePath: string): Promise<() => void> {
    return new Promise((res, rej) => {
      const httpRequest = new XMLHttpRequest();
      httpRequest.onload = () => {
        res(() => {
          eval(httpRequest.response);
        });
      };
      httpRequest.onerror = (error) => {
        rej(error);
      };
      httpRequest.open("GET", codePath);
      httpRequest.send();
    });
  }

  public static codeBlockWithFile(filePath: string): () => void {
    return () => {
      require(filePath).main();
    };
  }

  public static registerPlatformView(viewType: string, viewClass: typeof MPPlatformView) {
    ComponentFactory.components[viewType] = viewClass;
  }

  public initWithDebuggerServerAddr(debuggerServerAddr: string) {
    this.debugger = createDebugger(debuggerServerAddr, this);
  }

  public initWithCodeBlock(codeBlock: () => void) {
    this.codeBlock = codeBlock;
  }

  public start() {
    if (this.started) return;
    if (!this.codeBlock && !this.debugger) return;
    this.windowInfo.updateWindowInfo();
    this.listenViewport();
    MPEnv.platformGlobal();
    if (__MP_TARGET_BROWSER__) {
      MPEnv.platformGlobal().engineScope = this.mpJS.engineScope;
    } else {
      MPEnv.platformGlobal().JSON = JSON;
      MPEnv.platformGlobal().setTimeout = function (a: any, b: any) {
        var ret = setTimeout(a, b);
        return parseInt(ret as any);
      };
      MPEnv.platformGlobal().setInterval = function (a: any, b: any) {
        var ret = setInterval(a, b);
        return parseInt(ret as any);
      };
      MPEnv.platformGlobal().clearTimeout = function (v: any) {
        clearTimeout(v);
      };
      MPEnv.platformGlobal().clearInterval = function (v: any) {
        clearInterval(v);
      };
      MPEnv.platformGlobal().engineScope = this.mpJS.engineScope;
      MPEnv.platformGlobal().Object = Object;
      if (typeof wx !== "undefined") {
        MPEnv.platformGlobal().wx = wx;
        MPEnv.platformGlobal().uni = wx;
      }
      if (typeof swan !== "undefined") {
        MPEnv.platformGlobal().swan = swan;
        MPEnv.platformGlobal().uni = swan;
      }
      if (typeof tt !== "undefined") {
        MPEnv.platformGlobal().tt = tt;
        MPEnv.platformGlobal().uni = tt;
      }
    }
    if (this.debugger) {
      this.debugger.didConnectCallback = () => {
        this.windowInfo.updateWindowInfo();
      };
      this.debugger.start();
    }
    if (this.codeBlock) {
      this.mpJS.engineScope.onMessage = (message: string) => {
        this.didReceivedMessage(message);
      };
      this.mpJS.engineScope.onMapMessage = (message: any) => {
        this.didReceivedMessage(message, true);
      };
      (() => {
        this.codeBlock();
      })();
    }
    this.started = true;
  }

  sendMessage(message: string) {
    if (MPEnv.platformGlobal()?.mpDEBUG) {
      console.log(new Date().getTime(), "out", JSON.parse(message));
    }
    if (this.debugger) {
      this.debugger.sendMessage(message);
    } else {
      if (this.mpJS.engineScope.postMessage) {
        this.mpJS.engineScope.postMessage(message);
      } else {
        this.messageQueue.push(message);
      }
    }
  }

  flushQueueMessage() {
    if (this.debugger) {
      return;
    } else {
      if (this.mpJS.engineScope.postMessage) {
        this.messageQueue.forEach((it) => {
          this.mpJS.engineScope.postMessage(it);
        });
      }
    }
  }

  didReceivedMessage(message: string, isDartObject: boolean = false) {
    let decodedMessage = isDartObject ? wrapDartObject(message) : JSON.parse(message);
    if (!decodedMessage) return;
    if (MPEnv.platformGlobal()?.mpDEBUG) {
      console.log(new Date().getTime(), "in", decodedMessage);
    }
    if (decodedMessage.type === "ready") {
      this.flushQueueMessage();
    } else if (decodedMessage.type === "frame_data") {
      this.didReceivedFrameData(decodedMessage.message);
    } else if (decodedMessage.type === "diff_data") {
      this.didReceivedDiffData(decodedMessage.message);
    } else if (decodedMessage.type === "element_gc") {
      if (decodedMessage.message instanceof Array) {
        this.didReceivedElementGC(decodedMessage.message);
      }
    } else if (decodedMessage.type === "decode_drawable") {
      this.drawable.decodeDrawable(decodedMessage.message);
    } else if (decodedMessage.type === "custom_paint") {
      CustomPaint.didReceivedCustomPaintMessage(decodedMessage.message, this);
    } else if (decodedMessage.type === "route") {
      (this.app?.router ?? this.router)?.didReceivedRouteData(decodedMessage.message);
    } else if (decodedMessage.type === "mpjs") {
      this.didReceivedMPJS(decodedMessage);
    } else if (decodedMessage.type === "action:web_dialogs") {
      WebDialogs.receivedWebDialogsMessage(this, decodedMessage.message);
    } else if (decodedMessage.type === "scaffold") {
      this.didReceivedScaffold(decodedMessage.message);
    } else if (decodedMessage.type === "scroll_view") {
      this.didReceivedScrollView(decodedMessage.message);
    } else if (decodedMessage.type === "editable_text") {
      this.didReceivedEditableText(decodedMessage.message);
    } else if (decodedMessage.type === "rich_text" && decodedMessage.message?.event === "doMeasure") {
      TextMeasurer.didReceivedDoMeasureData(this, decodedMessage.message);
    } else if (decodedMessage.type === "rich_text" && decodedMessage.message?.event === "doMeasureTextPainter") {
      TextMeasurer.didReceivedDoMeasureTextPainter(this, decodedMessage.message);
    } else if (decodedMessage.type === "platform_view") {
      this.didReceivedPlatformView(decodedMessage.message);
    } else if (decodedMessage.type === "platform_channel") {
      this.platformChannelIO.didReceivedPlatformChannel(decodedMessage.message);
    }
  }

  didReceivedFrameData(frameData: { [key: string]: any }) {
    if (!frameData) return;
    const routeId = frameData.routeId;
    if (this.managedViews[routeId] === undefined) {
      if (!this.unmanagedViewFrameData[routeId]) {
        this.unmanagedViewFrameData[routeId] = [];
      }
      this.unmanagedViewFrameData[routeId].push(frameData);
      return;
    }
    if (typeof routeId === "number") {
      this.managedViews[routeId]?.didReceivedFrameData(frameData);
    }
  }

  didReceivedDiffData(frameData: { [key: string]: any }) {
    if (!frameData) return;
    (frameData.diffs as any[]).forEach((it) => {
      this.componentFactory.create(it, undefined!);
      this.componentFactory.createAncestors(it, undefined!);
    });
  }

  didReceivedElementGC(elements: number[]) {
    elements.forEach((it) => {
      this.componentFactory.cachedView[it]?.dispose();
      delete this.componentFactory.cachedElement[it];
      delete this.componentFactory.cachedView[it];
    });
  }

  didReceivedScaffold(message: any) {
    if (message.event === "onRefreshEnd") {
      let target = this.componentFactory.cachedView[message.target] as MPScaffold;
      if (target) {
        target.refreshEndResolver?.(undefined);
        target.refreshEndResolver = undefined;
      }
    } else if (message.event === "onWechatMiniProgramShareAppMessageResolve") {
      let target = this.componentFactory.cachedView[message.target] as MPScaffold;
      if (target) {
        target.onWechatMiniProgramShareAppMessageResolver?.(message.params);
        target.onWechatMiniProgramShareAppMessageResolver = undefined;
      }
    }
  }

  didReceivedScrollView(message: any) {
    if (message.event === "onRefreshEnd") {
      let target = this.componentFactory.cachedView[message.target] as CollectionView;
      if (target) {
        target.refreshEndResolver?.(undefined);
        target.refreshEndResolver = undefined;
      }
    } else if (message.event === "jumpTo") {
      let target = this.componentFactory.cachedView[message.target] as CollectionView;
      if (target) {
        target.jumpTo(message.value);
      }
    }
  }

  didReceivedEditableText(message: any) {
    if (message.event === "unfocus") {
      let target = this.componentFactory.cachedView[message.target] as EditableText;
      if (target) {
        target.contentElement?.blur();
      }
    } else if (message.event === "focus") {
      let target = this.componentFactory.cachedView[message.target] as EditableText;
      if (target) {
        target.contentElement?.focus();
      }
    }
  }

  didReceivedMPJS(decodedMessage: any) {
    this.mpJS.handleMessage(
      decodedMessage.message,
      (result) => {
        this.sendMessage(
          JSON.stringify({
            type: "mpjs",
            message: {
              requestId: decodedMessage.message.requestId,
              result: result,
            },
          })
        );
      },
      (funcId: string, args: any[]) => {
        this.sendMessage(
          JSON.stringify({
            type: "mpjs",
            message: {
              funcId: funcId,
              arguments: args,
            },
          })
        );
      }
    );
  }

  async didReceivedPlatformView(message: any) {
    if (message.event === "methodCall") {
      let target = this.componentFactory.cachedView[message.hashCode] as MPPlatformView;
      if (target) {
        const result = await target.onMethodCall(message.method, message.params);
        if (message.requireResult) {
          this.sendMessage(
            JSON.stringify({
              type: "platform_view",
              message: {
                event: "methodCallCallback",
                seqId: message.seqId,
                result: result,
              },
            })
          );
        }
      }
    } else if (message.event === "methodCallCallback") {
      const seqId = message["seqId"];
      if (typeof seqId === "string") {
        MPPlatformView.handleInvokeMethodCallback(seqId, message["result"]);
      }
    }
  }

  private listenViewport() {
    if (__MP_TARGET_BROWSER__) {
      window.addEventListener("resize", () => {
        this.sendViewportChangeEvent();
      });
    }
  }

  private sendViewportHandler: any;

  private sendViewportChangeEvent() {
    if (this.sendViewportHandler) {
      clearTimeout(this.sendViewportHandler);
    }
    this.sendViewportHandler = setTimeout(() => {
      this.windowInfo.updateWindowInfo();
      (this.app?.router as any)?.history.forEach((it: any) => {
        it.item.viewportChanged();
      });
      this.sendViewportHandler = undefined;
    }, 32);
  }
}

File: packages/mp_flutter_runtime/lib/engine.dart
part of './mp_flutter_runtime.dart';

abstract class MPDataReceiver {
  NavigatorState? getNavigator();
  BuildContext? getContext();
  void didReceivedFrameData(Map message);
}

class MPEngine {
  bool _started = false;
  bool get started => _started;
  final BuildContext flutterContext;
  final Map<int, MPDataReceiver> _managedViews = {};
  String? _jsCode;
  final List<String> _messageQueue = [];
  _MPDebugger? debugger;
  _MPKReader? _mpkReader;
  late _JSContext _jsContext;
  late _MPJS _mpjs;
  late _MPComponentFactory _componentFactory;
  late _MPRouter _router;
  late _TextMeasurer _textMeasurer;
  late _DrawableStore _drawableStore;
  late _MPPlatformChannelIO _platformChannelIO;
  var provider = MPProvider();

  MPEngine({required this.flutterContext}) {
    _componentFactory = _MPComponentFactory(engine: this);
    _router = _MPRouter(engine: this);
    _textMeasurer = _TextMeasurer(engine: this);
    _drawableStore = _DrawableStore(engine: this);
    _jsContext = _JSContext();
    _mpjs = _MPJS(engine: this);
    _platformChannelIO = _MPPlatformChannelIO(engine: this);
  }

  void initWithJSCode(String jsCode) {
    _jsCode = jsCode;
  }

  void initWithDebuggerServerAddr(String debuggerServerAddr) {
    debugger = _MPDebugger(engine: this, serverAddr: debuggerServerAddr);
  }

  void initWithMpkData(Uint8List mpkData) {
    _mpkReader = _MPKReader(mpkData);
    Uint8List? mainDataJSData = _mpkReader?.dataWithFilePath('main.dart.js');
    if (mainDataJSData != null) {
      _jsCode = utf8.decode(mainDataJSData);
    }
  }

  Future start() async {
    if (_started) return;
    if (_jsCode == null && debugger == null) return;
    updateWindowInfo();
    await _jsContext.createContext();
    await _setupJSContextEventChannel();
    await _setupDeferredLibraryLoader();
    await _MPJS.install(_jsContext);
    await _JSDeviceInfo.install(_jsContext, flutterContext);
    await _JSWXCompat.install(_jsContext);
    await _JSNetworkHttp.install(_jsContext, this);
    await _JSStorage.install(_jsContext, this);
    if (_jsCode != null) {
      await _jsContext.evaluateScript(_jsCode!);
    } else if (debugger != null) {
      debugger!.start();
    }
    for (final it in _messageQueue) {
      _jsContext.postMessage(it, '\$engine');
    }
    _messageQueue.clear();
    _started = true;
  }

  void stop() {
    debugger?.stop();
  }

  void clear() {
    _componentFactory.clear();
  }

  void updateWindowInfo() {
    _sendMessage({
      "type": "window_info",
      "message": {
        "window": {
          "width": MediaQuery.of(flutterContext).size.width,
          "height": MediaQuery.of(flutterContext).size.height,
          "padding": {
            "top": MediaQuery.of(flutterContext).padding.top,
            "bottom": MediaQuery.of(flutterContext).padding.bottom,
          }
        },
        "devicePixelRatio": MediaQuery.of(flutterContext).devicePixelRatio,
        "darkMode": MediaQuery.of(flutterContext).platformBrightness ==
            ui.Brightness.dark,
      },
    });
  }

  Future _setupJSContextEventChannel() async {
    _jsContext.addMessageListener((message, type) {
      if (type == '\$engine') {
        _didReceivedMessage(message);
      }
    });
    await _jsContext.evaluateScript('let self = globalThis');
    await _jsContext.evaluateScript('''
    globalThis.engineScope = {
      onMessage: function(message) {
        globalThis.postMessage(message, '\$engine');
      },
    };
    globalThis.onMessage = function(message, type) {
      if (type == '\$engine') {
        globalThis.engineScope.postMessage(message);
      }
    }
    ''');
  }

  Future _setupDeferredLibraryLoader() async {
    _jsContext.addMessageListener((fileName, type) async {
      if (type == '\$engine.dartDeferredLibraryLoader') {
        if (_mpkReader != null) {
          final data = _mpkReader!.dataWithFilePath(fileName);
          if (data != null) {
            final code = utf8.decode(data);
            try {
              await _jsContext.evaluateScript(code);
              await _jsContext.evaluateScript('''
              globalThis.dartDeferredLoaderPromiser['$fileName'].resolver(null);
              ''');
            } catch (e) {
              await _jsContext.evaluateScript('''
              globalThis.dartDeferredLoaderPromiser['$fileName'].rejector(null);
              ''');
            }
          }
        }
      }
    });
    await _jsContext.evaluateScript('''
    globalThis.dartDeferredLoaderPromiser = {};
    globalThis.dartDeferredLibraryLoader = function(fileName, resolver, rejector) {
      globalThis.dartDeferredLoaderPromiser[fileName] = {resolver: resolver, rejector: rejector};
      globalThis.postMessage(fileName, '\$engine.dartDeferredLibraryLoader');
    }
    ''');
  }

  void _didReceivedMessage(String message) {
    final decodedMessage = json.decode(message) as Map;
    String type = decodedMessage['type'];
    if (decodedMessage['message'] == null) return;
    switch (type) {
      case 'frame_data':
        _didReceivedFrameData(decodedMessage['message']);
        break;
      case 'diff_data':
        _didReceivedDiffData(decodedMessage['message']);
        break;
      case 'element_gc':
        _didReceivedElementGC(decodedMessage['message']);
        break;
      case 'decode_drawable':
        _drawableStore.decodeDrawable(decodedMessage['message']);
        break;
      case 'route':
        _router._didReceivedRouteData(decodedMessage['message']);
        break;
      case 'custom_paint':
        _CustomPaint._didReceivedCustomPaintMessage(
            decodedMessage['message'], this);
        break;
      case 'rich_text':
        if (decodedMessage['message']['event'] == 'doMeasure') {
          _textMeasurer._didReceivedDoMeasureData(decodedMessage['message']);
        } else if (decodedMessage['message']['event'] ==
            'doMeasureTextPainter') {
          _textMeasurer
              ._didReceivedDoMeasureTextPainer(decodedMessage['message']);
        }
        break;
      case 'action:web_dialogs':
        _WebDialogs.didReceivedWebDialogsMessage(
            decodedMessage['message'], this);
        break;
      case 'scroll_view':
        _didReceivedScrollView(decodedMessage['message']);
        break;
      case 'platform_view':
        MPPlatformView._didReceivedPlatformViewMessage(
            decodedMessage['message'], this);
        break;
      case 'mpjs':
        _mpjs._didReceivedMessage(decodedMessage['message']);
        break;
      case 'platform_channel':
        _platformChannelIO._didReceivedMessage(decodedMessage['message']);
        break;
      default:
    }
  }

  void _didReceivedFrameData(Map frameData) {
    int routeId = frameData['routeId'];
    MPDataReceiver? targetView = _managedViews[routeId];
    if (targetView != null) {
      targetView.didReceivedFrameData(frameData);
    }
  }

  void _didReceivedDiffData(Map frameData) {
    List diffs = frameData['diffs'];
    for (final data in diffs) {
      if (data is Map) {
        int? hashCode = data['hashCode'];
        if (hashCode != null) {
          _componentFactory._cacheViews[hashCode]?.updateData(data);
        }
      }
    }
  }

  void _didReceivedElementGC(List data) {
    for (final item in data) {
      if (item is int) {
        _componentFactory._cacheViews.remove(item);
      }
    }
  }

  void _didReceivedScrollView(Map data) {
    if (data["event"] == "onRefreshEnd") {
      final target = _RefresherManager.findCompleter(data["target"]);
      if (target != null) {
        target.complete();
      }
    } else if (data["event"] == "jumpTo") {
      final target = _ScrollControllerManager.findController(data["target"]);
      if (target != null) {
        target.jumpTo(data["value"]);
      }
    }
  }

  void _addManageView(int viewId, MPDataReceiver view) {
    _managedViews[viewId] = view;
  }

  void _sendMessage(Map mapMessage) {
    String message = json.encode(mapMessage);
    if (debugger != null) {
      debugger!.sendMessage(message);
    } else {
      if (_jsContext._contextRef == null) {
        _messageQueue.add(message);
        return;
      }
      _jsContext.postMessage(message, '\$engine');
    }
  }
}
